Implement instanced WebGL drawing calls (part of #20791)

nox opened the pull request:
This is half of #20599. The check for drawElements is a bit more complex to implement.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/21129)
<!-- Reviewable:end -->

nox committed:
Implement instanced WebGL drawing calls (part of #20791)

nox committed:
Correctly implement the vertex buffer checks in drawArrays

This is half of #20599. The check for drawElements is a bit more complex
to implement.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webglrenderingcontext.rs
 * @kichjang: components/script/dom/webglrenderingcontext.rs

nox commented:
r? @emilio 
@bors-servo try

bors-servo commented:
:hourglass: Trying commit c41bf6f041105684056678d24678b9a7185cd28b with merge 13b64fbd6f28bc1ae5a4fde561411f66f657eb7f...

Manishearth reviewed (approved):
code looks good but I don't quite understand the context here, may need a sanity check from emilio

nox committed:
Fix some outdated expectations

nox committed:
Pass more GL limits to the ANGLE shader compiler

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 578d3f2da152ccabce8362065dae00da0d591d32 with merge 0ecfaa4daef7ebc84a034a4e0d33eff1194c4c82...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4708)

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 3f2ed5ed2e5e085ac9541c4ceb7683467485b1c0 with merge 47828c5190f1a0684956d455dc8fc24a84905397...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9065)

nox committed:
Fix gl.getShaderSource and gl.getShaderInfoLog

It only returns null if there was an error, and the only error isn't
implemented yet.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit aa2aaad08fd7fbc73e38909674a2ef79302eca38 with merge bf197c8d5af57e2cf9f9fe491481a4b5d36d98f0...

nox committed:
Reenable .../ogles/GL/build/* (fixes #13165)

gl.getProgramInfoLog was implemented a while back.

nox committed:
Enable some tests marked as flaky

They don't seem flaky anymore and we have a tool to properly keep intermittent
results anyway now.

nox committed:
Return typed arrays from gl.getParameter (fixes #20655)

nox committed:
Implement gl.getParameter(gl.MAX_VIEWPORT_DIMS)

nox committed:
Return an empty typed array from gl.getParameter(gl.COMPRESSED_TEXTURE_FORMATS)

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -1155,6 +1157,140 @@ impl WebGLRenderingContext {
         }
     }
 
+    pub fn draw_arrays_instanced(
```
please add spec links?

nox replied:
Done.

nox commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 42e9a63 has been approved by `emilio`

<!-- @bors-servo r=emilio 42e9a6330fe65a56a90272041458e7e01ad23e86 -->

bors-servo commented:
:hourglass: Testing commit 42e9a6330fe65a56a90272041458e7e01ad23e86 with merge 25db2c6a708f0ee8cb2cbd02c81eb736c3402552...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9067)

CYBAI commented:
```
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/webgl/conformance-1.0.3/conformance/canvas/drawingbuffer-test.html", "line": 184495, "action": "test_result", "expected": "ERROR"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/webgl/conformance-1.0.3/conformance/canvas/drawingbuffer-static-canvas-test.html", "line": 193609, "action": "test_result", "expected": "ERROR"}
```

```
{"status": "FAIL", "group": "default", "message": "assert_true: testOES20Methods expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/more/unit.js:901:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1535:20\ntest@http://web-platform.test:8000/resources/testharness.js:545:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/more/unit.js:900:5\nrunTests@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/more/unit.js:189:5\ninitTests@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/more/unit.js:923:5\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/more/unit.js:948:7\n", "subtest": "WebGL test #0: testOES20Methods", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/more/conformance/methods.html", "line": 51085, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: context.getParameter(context.COMPRESSED_TEXTURE_FORMATS) is not an instance of Uint32Array expected true got false", "stack": "reportTestResultsToHarness/<@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1535:20\ntest@http://web-platform.test:8000/resources/testharness.js:545:9\nreportTestResultsToHarness@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7\ntestFailed@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5\nshouldBeType@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/resources/js-test-pre.js:469:9\n@http://web-platform.test:8000/_mozilla/webgl/conformance-1.0.3/conformance/state/gl-get-calls.html:85:5\n", "subtest": "WebGL test #20: context.getParameter(context.COMPRESSED_TEXTURE_FORMATS) is not an instance of Uint32Array", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/state/gl-get-calls.html", "line": 51928, "action": "test_result", "expected": "PASS"}
```

nox commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 4e6eea2 has been approved by `emilio`

<!-- @bors-servo r=emilio 4e6eea221a19b3e5a4b70f5954b3347bb6db02d1 -->

bors-servo commented:
:hourglass: Testing commit 4e6eea221a19b3e5a4b70f5954b3347bb6db02d1 with merge 41e105733c594a5660efa018cbb61a5ed67cb82d...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9070)

nox commented:
@bors-servo retry #21067 #20734

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/11853), [arm32](http://build.servo.org/builders/arm32/builds/12073), [arm64](http://build.servo.org/builders/arm64/builds/12121), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12160), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9068), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12211), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4627), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4648), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8950), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4713), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4647), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7558) are reusable. Rebuilding only [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9070), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8898)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/11853), [arm32](http://build.servo.org/builders/arm32/builds/12073), [arm64](http://build.servo.org/builders/arm64/builds/12121), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12160), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9068), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9071), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12211), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4627), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4648), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8899), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8950), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4713), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4647), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7558)
Approved by: emilio
Pushing 41e105733c594a5660efa018cbb61a5ed67cb82d to master...
